File: Data Structures & Algorithms/valid-tree/submission-1.py
from collections import deque, defaultdict
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        if len(edges) != n-1:
            return False
        visited = set()
        graph = defaultdict(list)
        for edge in edges:
            src = edge[0]
            dest = edge[1]
            graph[src].append(dest)
            graph[dest].append(src)
        queue = deque([0])
        while queue:
            node = queue.popleft()
            if node in visited:
                continue
            visited.add(node)
            for neighbor in graph[node]:
                if neighbor not in visited:
                    queue.append(neighbor)
        return len(visited) == n

File: Data Structures & Algorithms/count-connected-components/submission-1.java
class Solution {
    public int countComponents(int n, int[][] edges) {
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }
        for (int[] edge : edges) {
            int from = edge[0];
            int to = edge[1];
            graph.get(from).add(to);
            graph.get(to).add(from);
        }
        boolean[] visited = new boolean[n];
        int numComponents = 0;
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                dfs(visited, i, graph);
                numComponents++;
            }
        }
        return numComponents;
    }
    private void dfs(boolean[] visited, int node, List<List<Integer>> graph) {
        if (visited[node]) {
            return;
        }
        visited[node] = true;
        for (int neighbor: graph.get(node)) {
            dfs(visited, neighbor, graph);
        }
    }
}
